File: vripper-server/src/main/java/tn/mnlr/vripper/download/DownloadService.java
package tn.mnlr.vripper.download;

import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.mnlr.vripper.host.Host;
import tn.mnlr.vripper.jpa.domain.Image;
import tn.mnlr.vripper.jpa.domain.Post;
import tn.mnlr.vripper.jpa.domain.enums.Status;
import tn.mnlr.vripper.services.DataService;
import tn.mnlr.vripper.services.MetadataService;
import tn.mnlr.vripper.services.SettingsService;
import tn.mnlr.vripper.services.domain.Settings;

import javax.annotation.PostConstruct;
import java.util.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

@Service
@Slf4j
public class DownloadService {

  private final int MAX_POOL_SIZE = 12;

  private final SettingsService settingsService;
  private final DataService dataService;
  private final MetadataService metadataService;
  private final List<Host> hosts;

  private final Map<Host, AtomicInteger> threadCount = new HashMap<>();
  private final ExecutorService executor = Executors.newFixedThreadPool(MAX_POOL_SIZE);
  private final List<DownloadJob> running = new ArrayList<>();
  private final List<DownloadJob> pending = new ArrayList<>();

  private final Thread pollThread;

  @Autowired
  public DownloadService(
      SettingsService settingsService,
      DataService dataService,
      MetadataService metadataService,
      List<Host> hosts) {
    this.settingsService = settingsService;
    this.dataService = dataService;
    this.metadataService = metadataService;
    this.hosts = hosts;
    pollThread =
        new Thread(
            () -> {
              while (!Thread.interrupted()) {
                try {
                  synchronized (this) {
                    List<DownloadJob> accepted = new ArrayList<>();
                    List<DownloadJob> candidates = getCandidates(candidateCount());
                    candidates.forEach(
                        c -> {
                          if (canRun(c.getImage().getHost())) {
                            accepted.add(c);
                          }
                        });
                    pending.removeAll(accepted);
                    accepted.forEach(this::schedule);
                    accepted.clear();
                    this.wait();
                  }
                } catch (InterruptedException e) {
                  Thread.currentThread().interrupt();
                  break;
                }
              }
            },
            "Download scheduler thread");
  }

  @PostConstruct
  private void init() {
    pollThread.start();
  }

  public void destroy() throws Exception {
    log.info("Shutting down ExecutionService");
    pollThread.interrupt();
    executor.shutdown();
    dataService
        .findAllPosts()
        .forEach(
            p -> {
              log.debug(String.format("Stopping download jobs for %s", p));
              this.stopRunning(p.getPostId());
            });
    if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
      log.warn("Some jobs are still running!, forcing shutdown");
      executor.shutdownNow();
    }
  }

  private synchronized void stopRunning(@NonNull String postId) {
    List<DownloadJob> stopping = new ArrayList<>();
    Iterator<DownloadJob> iterator = running.iterator();
    while (iterator.hasNext()) {
      DownloadJob downloadJob = iterator.next();
      if (postId.equals(downloadJob.getPost().getPostId())) {
        downloadJob.stop();
        iterator.remove();
        stopping.add(downloadJob);
      }
    }

    while (!stopping.isEmpty()) {
      stopping.removeIf(DownloadJob::isFinished);
      try {
        Thread.sleep(100);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
      }
    }
  }

  public void stopAll(List<String> postIds) {
    stop(
        Objects.requireNonNullElseGet(
            postIds,
            () ->
                dataService.findAllPosts().stream()
                    .map(Post::getPostId)
                    .collect(Collectors.toList())));
  }

  public void restartAll(List<String> posIds) {
    restart(
        Objects.requireNonNullElseGet(
            posIds,
            () ->
                dataService.findAllPosts().stream()
                    .map(Post::getPostId)
                    .collect(Collectors.toList())));
  }

  private void restart(@NonNull List<String> postIds) {
    Map<Post, Collection<Image>> data = new HashMap<>();
    for (String postId : postIds) {
      if (isPending(postId)) {
        log.warn(
            String.format("Cannot restart, jobs are currently running for post id %s", postIds));
        continue;
      }
      List<Image> images = dataService.findByPostIdAndIsNotCompleted(postId);
      if (images.isEmpty()) {
        continue;
      }
      Post post = dataService.findPostByPostId(postId).orElseThrow();
      log.debug(String.format("Restarting %d jobs for post id %s", images.size(), postIds));
      data.put(post, images);
    }
    enqueue(data);
  }

  private synchronized boolean isPending(String postId) {
    return pending.stream().anyMatch(p -> p.getPost().getPostId().equals(postId));
  }

  private synchronized boolean isRunning(String postId) {
    return running.stream().anyMatch(p -> p.getPost().getPostId().equals(postId));
  }

  private synchronized void stop(List<String> postIds) {

    for (String postId : postIds) {
      final Post post = dataService.findPostByPostId(postId).orElseThrow();
      if (post == null) {
        continue;
      }
      pending.removeIf(p -> p.getPost().equals(post));
      stopRunning(postId);
      dataService.stopImagesByPostIdAndIsNotCompleted(postId);
      dataService.finishPost(post);
    }
    metadataService.stopFetchingMetadata(postIds);
  }

  private boolean canRun(Host host) {
    boolean canRun;
    int totalRunning = threadCount.values().stream().mapToInt(AtomicInteger::get).sum();
    canRun =
        threadCount.get(host).get() < settingsService.getSettings().getMaxThreads()
            && (settingsService.getSettings().getMaxTotalThreads() == 0
                ? totalRunning < MAX_POOL_SIZE
                : totalRunning < settingsService.getSettings().getMaxTotalThreads());
    if (canRun) {
      threadCount.get(host).incrementAndGet();
      return true;
    }
    return false;
  }

  private Map<Host, Integer> candidateCount() {
    HashMap<Host, Integer> map = new HashMap<>();
    hosts.forEach(
        h -> {
          AtomicInteger count = threadCount.get(h);
          if (count == null) {
            count = new AtomicInteger(0);
            threadCount.put(h, count);
          }
          map.put(h, settingsService.getSettings().getMaxThreads() - count.get());
        });
    return map;
  }

  private List<DownloadJob> getCandidates(Map<Host, Integer> candidateCount) {
    List<DownloadJob> candidates = new ArrayList<>();
    for (DownloadJob downloadJob : pending) {
      Host host = downloadJob.getImage().getHost();
      Integer maxPerHost = candidateCount.get(host);
      if (maxPerHost > 0) {
        candidates.add(downloadJob);
        candidateCount.put(host, maxPerHost - 1);
      }
    }
    return candidates;
  }

  public void enqueue(Map<Post, Collection<Image>> images) {
    synchronized (this) {
      for (Map.Entry<Post, Collection<Image>> entry : images.entrySet()) {
        entry.getKey().setStatus(Status.PENDING);
        dataService.updatePostStatus(entry.getKey().getStatus(), entry.getKey().getId());
        for (Image image : entry.getValue()) {
          log.debug(String.format("Enqueuing a job for %s", image.getUrl()));
          image.init();
          dataService.updateImageStatus(image.getStatus(), image.getId());
          dataService.updateImageCurrent(image.getCurrent(), image.getId());
          DownloadJob downloadJob =
              new DownloadJob(
                  entry.getKey(), image, (Settings) settingsService.getSettings().clone());
          pending.add(downloadJob);
        }
      }
      pending.sort(Comparator.comparing(e -> e.getPost().getAddedOn()));
      this.notify();
    }
  }

  private synchronized void schedule(DownloadJob downloadJob) {
    log.debug(String.format("Scheduling a job for %s", downloadJob.getImage().getUrl()));
    executor.execute(new DownloadJobWrapper(downloadJob));
    running.add(downloadJob);
  }

  public void afterJobFinish(DownloadJob downloadJob) {
    synchronized (this) {
      running.remove(downloadJob);
      threadCount.get(downloadJob.getImage().getHost()).decrementAndGet();
      if (!isPending(downloadJob.getPost().getPostId())
          && !isRunning(downloadJob.getPost().getPostId())) {
        dataService.finishPost(downloadJob.getPost());
      }
      this.notify();
    }
  }

  public int pendingCount() {
    return pending.size();
  }

  public int runningCount() {
    return running.size();
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/download/DownloadService.kt
package me.mnlr.vripper.download

import kotlinx.coroutines.*
import me.mnlr.vripper.delegate.LoggerDelegate
import me.mnlr.vripper.entities.Image
import me.mnlr.vripper.entities.LogEvent
import me.mnlr.vripper.entities.Post
import me.mnlr.vripper.entities.domain.Status
import me.mnlr.vripper.event.ErrorCountEvent
import me.mnlr.vripper.event.EventBus
import me.mnlr.vripper.event.QueueStateEvent
import me.mnlr.vripper.formatToString
import me.mnlr.vripper.host.Host
import me.mnlr.vripper.model.ErrorCount
import me.mnlr.vripper.model.QueueState
import me.mnlr.vripper.repositories.LogEventRepository
import me.mnlr.vripper.services.DataTransaction
import me.mnlr.vripper.services.DownloadSpeedService
import me.mnlr.vripper.services.RetryPolicyService
import me.mnlr.vripper.services.SettingsService
import net.jodah.failsafe.Failsafe
import net.jodah.failsafe.RetryPolicy
import java.util.concurrent.locks.ReentrantLock
import java.util.stream.Collectors
import kotlin.concurrent.withLock

class DownloadService(
    private val settingsService: SettingsService,
    private val dataTransaction: DataTransaction,
    private val retryPolicyService: RetryPolicyService,
    private val eventRepository: LogEventRepository,
    private val downloadSpeedService: DownloadSpeedService,
    private val eventBus: EventBus
) {
    private val maxPoolSize: Int = 12
    private val log by LoggerDelegate()

    // Class fields
    private val running: MutableMap<String, MutableList<ImageDownloadRunnable>> = mutableMapOf()
    private val pending: MutableMap<String, MutableList<ImageDownloadRunnable>> = mutableMapOf()
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun init() {
        coroutineScope.launch(Dispatchers.Default) {
            val accepted: MutableList<ImageDownloadRunnable> = mutableListOf()
            val candidates: MutableList<ImageDownloadRunnable> = mutableListOf()
            while (isActive) {
                lock.withLock {
                    candidates.addAll(getCandidates(candidateCount()))
                    candidates.forEach {
                        if (canRun(it.context.image.host)) {
                            accepted.add(it)
                            running[it.context.image.host]!!.add(it)
                            log.debug("${it.context.image.url} accepted to run")
                        }
                    }
                    accepted.forEach {
                        pending[it.context.image.host]?.remove(it)
                        scheduleForDownload(it)
                    }
                    accepted.clear()
                    candidates.clear()
                    try {
                        condition.await()
                    } catch (e: InterruptedException) {
                        Thread.currentThread().interrupt()
                    }
                }
            }
        }
    }

    fun destroy() {
        stop(dataTransaction.findAllPosts().map { it.postId })
    }

    fun stopAll(postIds: List<String>?) {
        if (postIds != null) {
            stop(postIds)
        } else {
            stop(dataTransaction.findAllPosts().map(Post::postId))
        }
    }

    fun restartAll(postIds: List<String> = listOf()) {
        if (postIds.isNotEmpty()) {
            restart(postIds)
        } else {
            restart(dataTransaction.findAllPosts().map(Post::postId))
        }
    }

    private fun restart(postIds: List<String>) {
        lock.withLock {
            val data: MutableMap<Post, Collection<Image>> = mutableMapOf()
            for (postId in postIds) {
                if (isPending(postId)) {
                    log.warn("Cannot restart, jobs are currently running for post id $postIds")
                    continue
                }
                val images: List<Image> =
                    dataTransaction.findByPostIdAndIsNotCompleted(postId)
                if (images.isEmpty()) {
                    continue
                }
                val post: Post =
                    dataTransaction.findPostsByPostId(postId).orElseThrow()
                log.debug("Restarting ${images.size} jobs for post id $postIds")
                post.status = Status.PENDING
                dataTransaction.update(post)
                data[post] = images
            }

            for ((_, images) in data) {
                for (image in images) {
                    log.debug("Enqueuing a job for ${image.url}")
                    with(image) {
                        this.status = Status.PENDING
                        this.current = 0
                    }
                    dataTransaction.update(image)
                    val imageDownloadRunnable = ImageDownloadRunnable(
                        image.id!!, settingsService.settings
                    )
                    pending.computeIfAbsent(
                        image.host
                    ) { mutableListOf() }
                    pending[image.host]!!.add(imageDownloadRunnable)
                }
            }
            condition.signal()
        }
    }

    private fun isPending(postId: String): Boolean {
        lock.withLock {
            return pending.values.flatten().any { it.context.image.postId == postId }
        }
    }

    private fun isRunning(postId: String): Boolean {
        lock.withLock {
            return running.values.flatten().any { it.context.image.postId == postId }
        }
    }

    private fun stop(postIds: List<String>) {
        lock.withLock {
            for (postId in postIds) {
                val post: Post =
                    dataTransaction.findPostsByPostId(postId).orElseThrow()
                pending.values.forEach { pending ->
                    pending.removeIf { it.context.image.postId == postId }
                }
                running.values.flatten()
                    .filter { p: ImageDownloadRunnable -> p.context.image.postId == postId }
                    .forEach { obj: ImageDownloadRunnable -> obj.stop() }
                dataTransaction.stopImagesByPostIdAndIsNotCompleted(postId)
                dataTransaction.finishPost(post)
            }
        }
    }

    private fun canRun(host: String): Boolean {
        val totalRunning = running.values.sumOf { it.size }
        return (running[host]!!.size < settingsService.settings.connectionSettings.maxThreads && if (settingsService.settings.connectionSettings.maxTotalThreads == 0) totalRunning < maxPoolSize else totalRunning < settingsService.settings.connectionSettings.maxTotalThreads)
    }

    private fun candidateCount(): Map<String, Int> {
        val map: MutableMap<String, Int> = mutableMapOf()
        Host.getHosts().forEach { host: String ->
            val imageDownloadRunnableList: List<ImageDownloadRunnable> = running.computeIfAbsent(
                host
            ) { mutableListOf() }
            val count: Int =
                settingsService.settings.connectionSettings.maxThreads - imageDownloadRunnableList.size
            log.debug("Download slots for $host: $count")
            map[host] = count
        }
        return map
    }

    private fun getCandidates(candidateCount: Map<String, Int>): List<ImageDownloadRunnable> {
        val hostIntegerMap: MutableMap<String, Int> = candidateCount.toMutableMap()
        val candidates: MutableList<ImageDownloadRunnable> = mutableListOf()
        hosts@ for (host in pending.keys) {
            val list: List<ImageDownloadRunnable> =
                pending[host]!!.sortedWith(Comparator.comparingInt<ImageDownloadRunnable> { it.context.post.rank }
                    .thenComparingInt { it.context.image.index })
            for (imageDownloadRunnable in list) {
                val count = hostIntegerMap[host] ?: 0
                if (count > 0) {
                    candidates.add(imageDownloadRunnable)
                    hostIntegerMap[host] = count - 1
                } else {
                    continue@hosts
                }
            }
        }
        if (log.isDebugEnabled) {
            val collect: Map<String, List<ImageDownloadRunnable>> =
                candidates.stream().collect(Collectors.groupingBy { it.context.image.host })
            collect.forEach {
                log.debug(
                    "Candidate download for ${it.key} ${it.value.size}/${candidateCount[it.key]}"
                )
            }
        }
        return candidates.sortedWith(Comparator.comparing { v: ImageDownloadRunnable -> v.context.post.rank })
    }

    private fun scheduleForDownload(imageDownloadRunnable: ImageDownloadRunnable) {
        log.debug("Scheduling a job for ${imageDownloadRunnable.context.image.url}")
        coroutineScope.launch {
            launch {
                eventBus.publishEvent(QueueStateEvent(QueueState(runningCount(), pendingCount())))
            }
            try {
                Failsafe.with<Any, RetryPolicy<Any>>(retryPolicyService.buildRetryPolicyForDownload())
                    .onFailure {
                        try {
                            eventRepository.save(
                                LogEvent(
                                    type = LogEvent.Type.DOWNLOAD,
                                    status = LogEvent.Status.ERROR,
                                    message = "Failed to download ${imageDownloadRunnable.context.image.url}\n ${it.failure.formatToString()}"
                                )
                            )
                        } catch (exp: Exception) {
                            log.error("Failed to save event", exp)
                        }
                        log.error(
                            "Failed to download ${imageDownloadRunnable.context.image.url} after ${it.attemptCount} tries",
                            it.failure
                        )
                        val image = imageDownloadRunnable.context.image
                        image.status = Status.ERROR
                        dataTransaction.update(image)
                        launch {
                            eventBus.publishEvent(ErrorCountEvent(ErrorCount(dataTransaction.countImagesInError())))
                        }
                    }.onComplete {
                        afterJobFinish(imageDownloadRunnable)
                        launch {
                            eventBus.publishEvent(
                                QueueStateEvent(
                                    QueueState(
                                        runningCount(),
                                        pendingCount()
                                    )
                                )
                            )
                        }
                        log.debug(
                            "Finished downloading ${imageDownloadRunnable.context.image.url}"
                        )
                    }.run(imageDownloadRunnable::run)
            } catch (ignored: Exception) {
            }
        }
    }

    private fun afterJobFinish(imageDownloadRunnable: ImageDownloadRunnable) {
        lock.withLock {
            val image = imageDownloadRunnable.context.image
            running[image.host]!!.remove(imageDownloadRunnable)
            if (!isPending(image.postId) && !isRunning(
                    image.postId
                )
            ) {
                dataTransaction.finishPost(imageDownloadRunnable.context.post)
            }
            condition.signal()
        }
    }

    fun pendingCount(): Int {
        return pending.values.sumOf { it.size }
    }

    fun runningCount(): Int {
        return running.values.sumOf { it.size }
    }
}